libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,686 @@
+    // @Range: 1 10
+    // @User: Standard
+    AP_GROUPINFO("OSC_SMAX", 5, AP_Quicktune, osc_smax, 5),
+
+    // @Param: QUIK_YAW_P_MAX
+    // @DisplayName: Quicktune Yaw P max
+    // @Description: Maximum value for yaw P gain
+    // @Range: 0.1 3
+    // @User: Standard
+    AP_GROUPINFO("YAW_P_MAX", 6, AP_Quicktune, yaw_p_max, 0.5),
+
+    // @Param: QUIK_YAW_D_MAX
+    // @DisplayName: Quicktune Yaw D max
+    // @Description: Maximum value for yaw D gain
+    // @Range: 0.001 1
+    // @User: Standard
+    AP_GROUPINFO("YAW_D_MAX", 7, AP_Quicktune, yaw_d_max, 0.01),
+
+    // @Param: QUIK_RP_PI_RATIO
+    // @DisplayName: Quicktune roll/pitch PI ratio
+    // @Description: Ratio between P and I gains for roll and pitch. Raise this to get a lower I gain
+    // @Range: 0.5 1.0
+    // @User: Standard
+    AP_GROUPINFO("RP_PI_RATIO", 8, AP_Quicktune, rp_pi_ratio, 1.0),
+
+    // @Param: QUIK_Y_PI_RATIO
+    // @DisplayName: Quicktune Yaw PI ratio
+    // @Description: Ratio between P and I gains for yaw. Raise this to get a lower I gain
+    // @Range: 0.5 20
+    // @User: Standard
+    AP_GROUPINFO("Y_PI_RATIO", 9, AP_Quicktune, y_pi_ratio, 10),
+
+    // @Param: QUIK_AUTO_FILTER
+    // @DisplayName: Quicktune auto filter enable
+    // @Description: When enabled the PID filter settings are automatically set based on INS_GYRO_FILTER
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Standard
+    AP_GROUPINFO("AUTO_FILTER", 10, AP_Quicktune, auto_filter, 1),
+
+    // @Param: QUIK_AUTO_SAVE
+    // @DisplayName: Quicktune auto save
+    // @Description: Number of seconds after completion of tune to auto-save. This is useful when using a 2 position switch for quicktune
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("AUTO_SAVE", 11, AP_Quicktune, auto_save, 0),
+
+    // @Param: QUIK_MAX_REDUCE
+    // @DisplayName: Quicktune maximum gain reduction
+    // @Description: This controls how much quicktune is allowed to lower gains from the original gains. If the vehicle already has a reasonable tune and is not oscillating then you can set this to zero to prevent gain reductions. The default of 20% is reasonable for most vehicles. Using a maximum gain reduction lowers the chance of an angle P oscillation happening if quicktune gets a false positive oscillation at a low gain, which can result in very low rate gains and a dangerous angle P oscillation.
+    // @Units: %
+    // @Range: 0 100
+    // @User: Standardd
+    AP_GROUPINFO("MAX_REDUCE", 12, AP_Quicktune, max_reduce, 20),
+
+    // @Param: QUIK_OPTIONS
+    // @DisplayName: Quicktune options
+    // @Description: Additional options. When the Two Position Switch option is enabled then a high switch position will start the tune, low will disable the tune. you should also set a QUIK_AUTO_SAVE time so that you will be able to save the tune.
+    // @Bitmask: 0:UseTwoPositionSwitch
+    // @User: Standard
+    AP_GROUPINFO("OPTIONS", 13, AP_Quicktune, options, 0),
```

Comment:
why not have the same behaviour as autotune? Simple on off and gains are saved on disarming without a mode change

Reply:
Right now we have 3 different autotune behaviors. Plane (fixed wing) saves when you switch out of Autotune, Quicktune saves on a switch and quad autotune saves on disarming and requires you don't turn it off in flight. Personally I find the quad autotune method the most confusing and have many times accidentally thrown away my tune by switching out of autotune before landing on a quad.

When you work with a quadplane you need to use all 3 different flavors of autotune making it easier to accidentally use the wrong "save" behavior depending on what tune you are running.

Personally I prefer the QuickTune method - you actively "save" the tune when you want to do a checkpoint or when you are good with it and done. Unfortunately if you are going to make quicktune a subclass of autotune, the simplest answer is to use the quad autotune behavior, but I'd love to see autotune and quicktune both work on a switch.

Even if we go with the current "save if still in autotune when disarm", at least that simplifies it down to 2 different behaviors instead of 3.

Reply:
One of the main reasons I like quicktune is because it doesn't force you to disarm in a specific mode. It is a workaround for when for some reason you can't use aux functions, but we're long past that. Having a switch is much more sensible, and much less likely to result in people accidentally wiping the tune they've just gotten.

Reply:
it is really common to change modes before saving to test the tune, I much prefer the behaviour we have in quicktune